azurelinuxagent/ga/agent_update.py

```diff
@@ -20,34 +20,101 @@ def get_agent_update_handler(protocol):
+        self._is_requested_version_update = True  # This is to track the current update type(requested version or self update)
+        self.persistent_data = AgentUpdateHandlerUpdateState()
 
     def __should_update_agent(self, requested_version):
         """
-        check to see if update is allowed once per (as specified in the conf.get_autoupdate_frequency())
-        return false when we don't allow updates.
+        requested version update:
+            update is allowed once per (as specified in the conf.get_autoupdate_frequency())
+            return false when we don't allow updates.
+        largest version update(self-update):
+            update is allowed once per (as specified in the conf.get_hotfix_upgrade_frequency() or conf.get_normal_upgrade_frequency())
+            return false when we don't allow updates.
         """
         now = datetime.datetime.now()
 
-        if self._last_attempted_update_time != datetime.datetime.min and self._last_attempted_update_version == requested_version:
-            next_attempt_time = self._last_attempted_update_time + datetime.timedelta(seconds=conf.get_autoupdate_frequency())
+        if self._is_requested_version_update:
+            if self.persistent_data.last_attempted_requested_version_update_time != datetime.datetime.min:
+                next_attempt_time = self.persistent_data.last_attempted_requested_version_update_time + datetime.timedelta(seconds=conf.get_autoupdate_frequency())
+            else:
+                next_attempt_time = now
+
+            if next_attempt_time > now:
+                return False
+            # The time limit elapsed for us to allow updates.
+            return True
         else:
-            next_attempt_time = now
+            next_hotfix_time, next_normal_time = self.__get_next_upgrade_times(now)
+            upgrade_type = self.__get_agent_upgrade_type(requested_version)
 
-        if next_attempt_time > now:
+            if next_hotfix_time > now and next_normal_time > now:
+                return False
+
+            if (upgrade_type == AgentUpgradeType.Hotfix and next_hotfix_time <= now) or (
+                    upgrade_type == AgentUpgradeType.Normal and next_normal_time <= now):
+                return True
             return False
-        # The time limit elapsed for us to allow updates.
-        return True
+
+    def __update_last_attempt_update_times(self):
+        now = datetime.datetime.now()
+        if self._is_requested_version_update:
+            self.persistent_data.last_attempted_requested_version_update_time = now
+        else:
+            self.persistent_data.last_attempted_normal_update_time = now
+            self.persistent_data.last_attempted_hotfix_update_time = now
+
+    @staticmethod
+    def __get_agent_upgrade_type(requested_version):
+        # We follow semantic versioning for the agent, if <Major>.<Minor> is same, then <Patch>.<Build> has changed.
```

Comment:
I just noticed that we have this bug in the develop branch as well. A change in <Patch> should also be treated as a Normal release, it indicates that the release has only bug fixes. The release should be hotfix only when the <Build> changes.

Reply:
sorry Norberto, I didn't get you. Can you point me with examples if you could?

Reply:
@nagworld9 I was just trying to point out that this is not your bug, we have the same issue in the develop branch.

If the current version is 1.1.1.1

* 2.0.0.0 would indicate a breaking change
* 1.2.0.0 would indicate a new feature
* 1.1.2.0 would indicate a bug fix
* 1.1.1.2 would indicate a hotfix/retake

The first 3 should be normal releases.